data_steward/cdr_cleaner/cleaning_rules/drop_participants_without_ppi_or_ehr.py

```diff
@@ -0,0 +1,114 @@
+
+BASICS_MODULE_CONCEPT_ID = 1586134
+
+SELECT_PERSON_WITH_BASICS_OR_EHR = """
+SELECT {fields}
+FROM `{project}.{dataset}.person` person
+LEFT JOIN (
+  SELECT DISTINCT o.person_id
+  FROM `{project}.{dataset}.concept_ancestor`
+  INNER JOIN `{project}.{dataset}.observation` o ON observation_concept_id = descendant_concept_id
+  INNER JOIN `{project}.{dataset}.concept` d ON d.concept_id = descendant_concept_id
+  WHERE ancestor_concept_id = {basics_module_concept_id}) basics
+ON
+  person.person_id = basics.person_id
+{ehr_joins}
+WHERE
+  basics.person_id IS NOT NULL
+  OR ({has_ehr_predicate})
+"""
+
+# Template for a JOIN against the person table against a table which may contain
+# EHR data.
+JOIN_EHR_PERSON_IDS_TEMPLATE = """
+LEFT JOIN (
+  SELECT DISTINCT v.person_id AS person_id
+  FROM `{project}.{dataset}.{table}` t
+  LEFT JOIN `{project}.{dataset}._mapping_{table}` m ON at.{id_column} = m.{id_column}
+  # The source HPO is either the "rdr", or a site ID; we only want to capture sites here.
+  WHERE m.src_hpo_id != "rdr"
+) {table}_ehr
+ON
+  person.person_id = {table}_ehr.person_id
+"""
+
+def get_queries(project=None, dataset=None):
+    """
+    Return a list of queries to remove data-poor participant rows.
+
+    The removal criteria is for participants is as follows:
+    1. They have not completed "The Basics" PPI module, via the RDR
+    2. They do not have any EHR data
+
+    These participants are not particularly useful for analysis, so remove them
+    here.
+
+    :return:  A list of string queries that can be executed to delete data-poor
+        participants and corresponding rows from the dataset.
+    """
+    field_names = ['entry.' + field['name'] for field in resources.fields_for('person')]
+    fields = ', '.join(field_names)
+
+    ehr_joins = ""
+    for t in common.MAPPED_CLINICAL_DATA_TABLES:
+        ehr_joins += JOIN_EHR_PERSON_IDS_TEMPLATE.format(
+            project=project, dataset=dataset, table=t,
+            id_column=resources.get_domain_id_field(t))
+    has_ehr_predicate = ' OR '.join(
+        ['{}.person_id IS NOT NULL'.format(t) for t in common.MAPPED_CLINICAL_DATA_TABLES])
+
+    delete_query = SELECT_PERSON_WITH_BASICS_OR_EHR.format(
+        project=project, dataset=dataset, fields=fields,
+        ehr_joins=ehr_joins, has_ehr_predicate=has_ehr_predicate,
+        basics_module_concept_id=BASICS_MODULE_CONCEPT_ID
+    )
+
+    # drop from the person table, then delete all corresponding data for the now missing persons
+    return [{
+        clean_consts.QUERY: delete_query,
+        clean_consts.DESTINATION_TABLE: 'person',
+        clean_consts.DESTINATION_DATASET: dataset,
+        clean_consts.DISPOSITION: bq_consts.WRITE_TRUNCATE,
+    }] + drop_rows_for_missing_persons.get_queries(project, dataset)
+
+
+if __name__ == '__main__':
+    import cdr_cleaner.clean_cdr_engine as clean_engine
+
+    parser = argparse.ArgumentParser(formatter_class=argparse.RawDescriptionHelpFormatter)
```

Comment:
@calbach I get unresolved reference `argparse` here- is an import missing?

Reply:
Added. Interestingly when I try to execute this against my bq test dataset I see:
```
Traceback (most recent call last):
  File "/usr/local/google/home/calbach/aou/curation/data_steward/cdr_cleaner/clean_cdr_engine.py", line 72, in clean_dataset
    batch=batch)
  File "/usr/local/google/home/calbach/aou/curation/data_steward/bq_utils.py", line 427, in query
    return bq_service.jobs().insert(projectId=app_id, body=job_body).execute(num_retries=retry_count)
  File "/usr/local/google/home/calbach/.virtualenvs/curation/lib/python3.7/site-packages/oauth2client/_helpers.py", line 133, in positional_wrapper
    return wrapped(*args, **kwargs)
  File "/usr/local/google/home/calbach/.virtualenvs/curation/lib/python3.7/site-packages/googleapiclient/http.py", line 842, in execute
    raise HttpError(resp, content, uri=self.uri)
googleapiclient.errors.HttpError: <HttpError 400 when requesting https://bigquery.googleapis.com/bigquery/v2/projects/aou-res-curation-test/jobs?alt=json returned "Required parameter is missing">

```

Reply:
I think this is related to a missing environment variable (guessing maybe `GOOGLE_CLOUD_PROJECT` or `BIGQUERY_DATASET_ID`)?

Reply:
I resolved this by iterating on my test. I believe GOOGLE_CLOUD_PROJECT was missing in this case, but there were subsequent failures due to syntax issues in the SQL which I don't think were reported when I ran this. I think using a standard BigQuery library will help immensely with this kind of issue.